pennylane/qnode_new/qnode.py

```diff
@@ -0,0 +1,935 @@
+                    pos.append(p)
+        return tuple(pos), kwargs
+
+
+    def _check_args(self, args, kwargs):
+        """NEW: Validate the quantum function arguments, apply defaults.
+
+        Here we apply all the default values for the parameters found in the signature of
+        :attr:`QNode.func`, and check if any arguments are invalid or missing.
+
+        Args:
+            args (tuple[Any]): positional (differentiable) arguments
+            kwargs (dict[str, Any]): auxiliary arguments
+
+        Returns:
+            tuple[Any], dict[str, Any]: positional arguments, auxiliary arguments
+        """
+        # pylint: disable=too-many-branches
+        # TODO maybe disallow *args in qfunc signature entirely? it complicates things and may be confusing
+
+        forbidden_kinds = (inspect.Parameter.POSITIONAL_ONLY,
+                           inspect.Parameter.VAR_POSITIONAL,
+                           inspect.Parameter.VAR_KEYWORD)
+
+        # check the validity of kwargs items
+        for name in kwargs:
+            s = self.func.sig.get(name)
+            if s is None:
+                if self.func.var_keyword:
+                    continue  # unknown parameter, but **kwargs will take it
+                raise QuantumFunctionError("Unknown quantum function parameter '{}'.".format(name))
+            if s.par.kind in forbidden_kinds:
+                raise QuantumFunctionError("Quantum function parameter '{}' cannot be given using the keyword syntax.".format(name))
+
+        n_args = len(args)  # number of positional args given
+        if n_args < self.func.n_pos or (n_args > self.func.n_pos and not self.func.var_pos):
+            raise QuantumFunctionError("Quantum function takes {} positional parameters, {} given.".format(self.func.n_pos, n_args))
+
+        # apply defaults, check for missing parameters
+        pos = list(args)
+        for name, s in self.func.sig.items():
+            default = s.par.default
+            if s.idx < n_args:
+                # positional
+                if pos[s.idx] == _MARKER:
+                    # missing
+                    if default != inspect.Parameter.empty:
+                        pos[s.idx] = default
+                    else:
+                        raise QuantumFunctionError("Quantum function positional parameter '{}' missing.".format(name))
+            else:
+                # keyword-only
+                # can we find the parameter in kwargs?
+                p = kwargs.get(name, _MARKER)  # None may be a valid parameter value, hence _MARKER
+                if p is _MARKER and s.par.kind not in forbidden_kinds:
+                    # missing
+                    if default != inspect.Parameter.empty:
+                        kwargs[name] = default
+                    else:
+                        raise QuantumFunctionError("Quantum function keyword-only parameter '{}' missing.".format(name))
+
+        return tuple(pos), kwargs
```

Comment:
This is a very complicated function - solely because of `*args`?

Reply:
This will require a lot of testing, there are a lot of edge cases

Reply:
Not just for *args, but disallowing it would simplify the function a bit. I already have tests for all the errors, but maybe there could be additional edge cases.